### PR TITLE
NeMo TTS collection

### PR DESCRIPTION
**Pull request by blisc**
- [x] Add neural modules for tacotron 2
- [x] Add neural modules for waveglow
- [x] Add training script for tacotron 2
- [x] Add training script for waveglow
- [x] Add inference script for TTS
- [x] Add unittests for TTS
- [x] Docstrings and comments for TTS code
- ~[ ] Doc pages for TTS~
- ~[ ] Train tacotron 2~
- ~[ ] Train waveglow~

### CODE REVIEW

**Comment by blisc** [comment]
@chiphuyen @tkornuta-nvidia I would appreciate any feedback on the code

File: collections/nemo_tts/nemo_tts/data_layers.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2019 NVIDIA Corporation
+import torch
+try:
+    from apex import amp
+except AttributeError:
```
Comment (okuchaiev) [inline collections/nemo_tts/nemo_tts/data_layers.py]: could you please change this to: "except (AttributeError, ModuleNotFoundError) as e:"

File: collections/nemo_tts/nemo_tts/parts/layers.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2019 NVIDIA Corporation
+import torch
+
+
+class LinearNorm(torch.nn.Module):
```
Comment (okuchaiev) [inline collections/nemo_tts/nemo_tts/parts/layers.py]: why call this LinearNorm, isn't this just "Linear"?

File: collections/nemo_tts/nemo_tts/parts/layers.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    def __init__(self, in_dim, out_dim, bias=True, w_init_gain='linear'):
```
Comment (okuchaiev) [inline collections/nemo_tts/nemo_tts/parts/layers.py]: should bias=True by default?

File: collections/nemo_tts/nemo_tts/parts/layers.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        torch.nn.init.xavier_uniform_(
+            self.linear_layer.weight,
+            gain=torch.nn.init.calculate_gain(w_init_gain))
+
+    def forward(self, x):
+        return self.linear_layer(x)
+
+
+class ConvNorm(torch.nn.Module):
```
Comment (okuchaiev) [inline collections/nemo_tts/nemo_tts/parts/layers.py]: Same thing - why "Norm" suffix?

File: collections/nemo_tts/nemo_tts/parts/layers.py (hunk 3 above)
Comment (okuchaiev) [reply collections/nemo_tts/nemo_tts/parts/layers.py]: I think we should do bias=False by default

File: collections/nemo_tts/nemo_tts/data_layers.py (hunk 1 above)
Comment (blisc) [reply collections/nemo_tts/nemo_tts/data_layers.py]: amp is no longer used in this file and will be removed

File: collections/nemo_tts/nemo_tts/parts/layers.py (hunk 2 above)
Comment (blisc) [reply collections/nemo_tts/nemo_tts/parts/layers.py]: TBH, I have no idea, I took this from Deep Learning Examples. I think for future releases, I will be moving away from reusing code from Deep Learning Examples, and just git submodule the nvidia tacotron 2 and waveglow repos.

File: collections/nemo_tts/nemo_tts/tacotron2_modules.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+                                       1: AxisType(TimeTag)}),
+            "alignments": NeuralType({0: AxisType(BatchTag),
+                                      1: AxisType(TimeTag),
+                                      2: AxisType(TimeTag)})
+        }
+        return input_ports, output_ports
+
+    def __init__(
+            self,
+            n_mel_channels,
```
[the remark this answers is not in the document]
Comment (blisc) [reply collections/nemo_tts/nemo_tts/tacotron2_modules.py]: Done.

File: collections/nemo_tts/nemo_tts/tacotron2_modules.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+            "mel_len": NeuralType({0: AxisType(BatchTag)})
+        }
+        return input_ports, output_ports
+
+    def __str__(self):
+        return "Tacotron2Decoder"
+
+    def __init__(
+            self,
+            n_mel_channels,
```
[the remark this answers is not in the document]
Comment (blisc) [reply collections/nemo_tts/nemo_tts/tacotron2_modules.py]: Done

File: collections/nemo_tts/nemo_tts/tacotron2_modules.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+        output_ports = {
+            "mel_output": NeuralType({0: AxisType(BatchTag),
+                                      1: AxisType(MelSpectrogramSignalTag),
+                                      2: AxisType(TimeTag)}),
+        }
+        return input_ports, output_ports
+
+    def __init__(
+            self,
+            n_mel_channels,
```
[the remark this answers is not in the document]
Comment (blisc) [reply collections/nemo_tts/nemo_tts/tacotron2_modules.py]: Done

File: collections/nemo_tts/nemo_tts/waveglow_modules.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            "audio": NeuralType({0: AxisType(BatchTag),
+                                 1: AxisType(TimeTag)}),
+            "log_s_list": NeuralType(),
+            "log_det_W_list": NeuralType(),
+        }
+        return input_ports, output_ports
+
+    def __init__(
+            self,
+            n_mel_channels=80,
```
[the remark this answers is not in the document]
Comment (blisc) [reply collections/nemo_tts/nemo_tts/waveglow_modules.py]: Done

File: collections/nemo_tts/setup.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
+
+setuptools.setup(
+    name="nemo_tts",
+    version="0.9",
+    author="NVIDIA",
+    author_email="nemo-toolkit@nvidia.com",
+    description="Collection of Neural Modules for Speech Recognition",
```
[the remark this answers is not in the document]
Comment (blisc) [reply collections/nemo_tts/setup.py]: Done

File: collections/nemo_tts/nemo_tts/data_layers.py (hunk 1 above)
Comment (okuchaiev) [reply collections/nemo_tts/nemo_tts/data_layers.py]: can you please remove this try except block then?

**Pull request merged**